### PR TITLE
Add deepcopy to setting data

### LINKED ISSUE #114: DataContainers only stores a deep copy of the data when !data_are_columns
We currently have

```
struct DataContainer{FT <: Real}
     #stored data, each piece of data is a column [data dimension × number samples]
     stored_data::AbstractMatrix{FT}
     #constructor with 2D arrays
     function DataContainer(stored_data::AbstractMatrix{FT}; data_are_columns = true) where {FT <: Real}

         if data_are_columns
             new{FT}(stored_data)
         else
             new{FT}(permutedims(stored_data, (2, 1)))
         end
     end
 end
```
The behavior for this struct is as follows,

```
julia > u = rand(5,3)

output > 5×3 Matrix{Float64}:
 0.553978   0.100353  0.455385
 0.393232   0.138405  0.74255
 0.0660367  0.174017  0.401449
 0.669909   0.231632  0.249801
 0.113858   0.4794    0.315958

julia > data1 = DataContainer(u)
julia> data2 = DataContainer(u, data_are_columns = false)

julia> get_data(data1)

output > 5×3 Matrix{Float64}:
 0.273102   0.24338   0.728426
 0.0066842  0.320747  0.562981
 0.420169   0.521042  0.620609
 0.431041   0.177357  0.631251
 0.284572   0.120433  0.876362

julia> u[1,1] = -5

julia> get_data(data1)

output > 5×3 Matrix{Float64}:
 -5.0        0.24338   0.728426
  0.0066842  0.320747  0.562981
  0.420169   0.521042  0.620609
  0.431041   0.177357  0.631251
  0.284572   0.120433  0.876362

# With data_are_columns, the memory is shared.

julia> get_data(data2)

3×5 Matrix{Float64}:
  0.273102  0.0066842  0.420169  0.431041  0.284572
  0.24338   0.320747   0.521042  0.177357  0.120433
  0.728426  0.562981   0.620609  0.631251  0.876362

# With !data_are_columns, the memory is not shared.

```

### PR DESCRIPTION
**Pull request by odunbar**
## Purpose
Resolves #114 

## In this PR
- deepcopy in `DataContainers` constructor
- tests to check copy on set and get.

### CODE REVIEW

File: src/DataContainers.jl [hunk 1]
```diff
@@ -19,11 +19,10 @@ struct DataContainer{FT <: Real}
     stored_data::AbstractMatrix{FT}
     #constructor with 2D arrays
     function DataContainer(stored_data::AbstractMatrix{FT}; data_are_columns = true) where {FT <: Real}
-
         if data_are_columns
-            new{FT}(stored_data)
+            new{FT}(deepcopy(stored_data))
         else
-            new{FT}(permutedims(stored_data, (2, 1)))
+            new{FT}(permutedims(deepcopy(stored_data), (2, 1)))
```
Comment (ilopezgp) [inline src/DataContainers.jl:25]: This one is not necessary, permutedims already performs a deep copy before transposing.

File: test/DataContainers/runtests.jl [hunk 2]
```diff
@@ -39,4 +39,11 @@ using EnsembleKalmanProcesses.DataContainers
     @test get_inputs(iopairs) == parameter_samples
     @test get_outputs(iopairs) == data_samples
 
+    # test deep-copying
+    parameter_samples[1, 1] += 1
+    @test !isequal(get_data(idata), parameter_samples)
+    retrieved_samples = get_data(idata)
+    retrieved_samples[1, 1] += 1
+    @test !isequal(retrieved_samples, get_data(idata))
```
Comment (ilopezgp) [inline test/DataContainers/runtests.jl:47]: Could you add similar tests for `idata_T`?

File: src/DataContainers.jl (hunk 1 above)
Comment (odunbar) [reply src/DataContainers.jl:25]: I thought it would be clearer this way, I'll add a comment instead

File: test/DataContainers/runtests.jl (hunk 2 above)
Comment (odunbar) [reply test/DataContainers/runtests.jl:47]: done

**Review by ilopezgp — APPROVED** [review]
LGTM, thank you!

**Pull request merged**